sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt

```diff
@@ -0,0 +1,331 @@
+            annotations = annotations,
+            creationDate = dateTimeFormatter.now(),
+            dataKey = cryptoService.generateGCKey().blockingGet()
+        )
+    }
+
+    private fun fetchCommonKey(): Pair<GCKey, String> {
+        return Pair(
+            cryptoService.fetchCurrentCommonKey(),
+            cryptoService.currentCommonKeyId
+        )
+    }
+
+    private fun encryptKey(
+        commonKey: GCKey,
+        mode: KeyType,
+        key: GCKey
+    ): EncryptedKey {
+        return cryptoService.encryptSymmetricKey(
+            commonKey,
+            mode,
+            key
+        ).blockingGet() as EncryptedKey
+    }
+
+    private fun encryptKeys(
+        commonKey: GCKey,
+        dataKey: GCKey,
+        attachmentKey: GCKey?
+    ): Pair<EncryptedKey, EncryptedKey?> {
+        val encryptedDataKey = encryptKey(commonKey, KeyType.DATA_KEY, dataKey)
+
+        return if (attachmentKey is GCKey) {
+            Pair(
+                encryptedDataKey,
+                encryptKey(commonKey, KeyType.ATTACHMENT_KEY, attachmentKey)
+            )
+        } else {
+            Pair(encryptedDataKey, null)
+        }
+    }
+
+    override fun <T : Any> encrypt(
+        decryptedRecord: NetworkModelContract.DecryptedBaseRecord<T>
+    ): NetworkModelContract.EncryptedRecord {
+        val dataKey = decryptedRecord.dataKey
+        val (commonKey, commonKeyId) = fetchCommonKey()
+        val (encryptedDataKey, encryptedAttachmentKey) = encryptKeys(
+            commonKey,
+            dataKey,
+            decryptedRecord.attachmentsKey
+        )
+
+        return EncryptedRecord(
+            commonKeyId,
+            decryptedRecord.identifier,
+            tagEncryptionService.encryptTagsAndAnnotations(
+                decryptedRecord.tags,
+                decryptedRecord.annotations
+            ),
+            fhirService._encryptResource(dataKey, decryptedRecord.resource),
+            decryptedRecord.customCreationDate,
+            encryptedDataKey,
+            encryptedAttachmentKey,
+            decryptedRecord.modelVersion
+        )
+    }
+
+    private fun validateRecord(encryptedResource: String, version: Int) {
+        if (encryptedResource.isEmpty()) {
+            throw CoreRuntimeException.InternalFailure()
+        }
+
+        if (!modelVersion.isModelVersionSupported(version)) {
+            throw DataValidationException.ModelVersionNotSupported("Please update SDK to latest version!")
+        }
+    }
+
+    private fun getCommonKey(commonKeyId: String, userId: String): GCKey {
+        val commonKeyStored = cryptoService.hasCommonKey(commonKeyId)
+        return if (commonKeyStored) {
+            cryptoService.getCommonKeyById(commonKeyId)
+        } else {
+            // TODO: This should be in a different Service
+            val commonKeyResponse = apiService.fetchCommonKey(
+                alias,
+                userId,
+                commonKeyId
+            ).blockingGet()
+
```

Comment:
I see the TODO :) we created a `CommonKeyService` to wrap up all the CommonKey shenanigans

Reply:
Indeed...the whole key stuff should be encapsulated somewhere else...but to do it right now cause more trouble than it helps. However this will happen once Koin (DI) is in place.